writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
-	// the file are all required to have a continuous countdown. The count can
-	// start at any value, and then it will count down for the whole page,
-	// underflowing if required.
-	//
-	// The first 80 bytes must be 10 8-byte values that count down and underflow.
-	// The next 160 bytes must be 20 8-byte values that count down and underflow.
-	// The next 240 bytes must be 30 8-byte values that count down and underflow.
-	// ...
-	//
-	// Different parts of this file can easily be edited in parallel
-	// transactions, while at the same time there are clear consistency
-	// requirements.
-	//
-	// '0' is used as a special starting value to indicate that the whole
-	// splotch should be empty.
-	splotchFile *os.File
-
-	wal *WAL
-}
+		// ocs is the checksum of the file that became obsolete after changing a
+		// number. It needs to be removed when the update is applied.
+		ocs checksum
 
-// addCount will increment every counter in the array, and then append a '1',
-// but using the wal to maintain consistency on disk.
-func (ca *countdownArray) addCount() error {
-	// Increment the count in memory, creating a list of updates as we go.
-	var updates []Update
-	for i := 0; i < len(ca.countdown); i++ {
-		ca.countdown[i]++
-		updates = append(updates, Update{
-			Name:    addCountName,
-			Version: "1.0.0",
-			Instructions: addCountUpdate{
-				index:    uint64(i),
-				newCount: ca.countdown[i],
-			}.marshal(),
-		})
-	}
+		// ncs is the new checkum of the silo and the corresponding file
+		ncs checksum
 
-	// Corner case - if there are no updates, because this is the first time we
-	// add to the count, just write out the file piece.
-	if len(updates) == 0 {
-		writeBytes := make([]byte, 8)
-		_, err := ca.file.WriteAt(writeBytes, 8*int64(len(ca.countdown)))
-		if err != nil {
-			return err
-		}
-		err = ca.file.Sync()
-		if err != nil {
-			return err
-		}
-		ca.countdown = append(ca.countdown, 0)
-		return nil
+		// ncso is the offset of the new checksum
+		ncso int64
 	}
+)
 
-	// Create the WAL transaction.
-	tx, err := ca.wal.NewTransaction(updates)
-	if err != nil {
-		return err
+// newSilo creates a new silo of a certain length at s specific offset in the
+// file
+func newSilo(offset int64, length int, deps dependencies, f file) *silo {
+	if length == 0 {
+		panic("numbers shouldn't be empty")
 	}
-	// Perform the setup write on the file.
-	writeBytes := make([]byte, 8)
-	_, err = ca.file.WriteAt(writeBytes, 8*int64(len(ca.countdown)))
-	if err != nil {
-		return err
-	}
-	err = ca.file.Sync()
-	if err != nil {
-		return err
-	}
-	// Signal completed setup and then wait for the commitment to finish.
-	errChan := tx.SignalSetupComplete()
-	err = <-errChan
-	if err != nil {
-		return err
-	}
-	// Apply the updates.
-	ca.countdown = append(ca.countdown, 0)
-	err = ca.applyUpdates(updates)
-	if err != nil {
-		return err
+
+	return &silo{
+		offset:  offset,
+		numbers: make([]uint32, length, length),
```

Comment:
Would it cause a problem with the test if the length is set randomly for each silo? Does the recovery function expect each offset to have an explicit length?

Reply:
I guess it doesn't expect an explicit length but it is easier to estimate the ram usage of the test this way. I don't think the test would profit from random silo lengths.